Make HivePrivilege a class to better reflect grant option

amrutagokhale opened the pull request:
This is a follow-up on adding support for grant. It re-factors the hive privileges code. 

The current implementation of Hive privileges represents grant option as another constant, `GRANT`, in `HivePrivilege` enum type. This representation doesn't reflect the specific privilege for which grant option was given. In the new implementation, hive privilege is represented as a `HivePrivilegeInfo` object which has two members -- 
- an enum type that represents the privilege
- a boolean `grantOption` which represents the grant option for the specific privilege in enum type

@dain, can you please take a look?

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrivilegeInfo.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    private final HivePrivilege hivePrivilege;
+    private boolean grantOption;
```
Make this final

amrutagokhale replied:
done

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HivePrivilegeInfo.java:
```diff
@@ -0,0 +1,97 @@
+        return grantOption;
+    }
+
+    public void setGrantOption(boolean grantOption)
```
This doesn't seem to be used.  If we do need something like this in the future, name it `withGrantOption` and return an new instance (so the call can be immutable)

amrutagokhale replied:
Removed it.

amrutagokhale replied:
I realized that we do need it later (for revoke, for ex.). Added a new method.

amrutagokhale commented:
Updated.

amrutagokhale commented:
I looked at the Travis build failure. It looks like an intermittent failure and is unrelated to my changes.

amrutagokhale commented:
Rebased and squashed commits. Travis build passed.

amrutagokhale commented:
@dain rebased again.

haozhun commented:
I'll integrate this pull request to #4878 today, where I replaced all entities in the metastore interface of Presto to use custom entity classes instead of hive.metastore.api thrift classes. Closing for now.

haozhun commented:
After reading more closely, I decided that we'll merge this first (before #4878) instead of incorporating it into #4878. Sorry about it.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java:
```diff
@@ -189,24 +190,36 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
+    private boolean getGrantOptionForPrivilege(Identity identity, Privilege privilege, SchemaTableName tableName)
+    {
+        return metastore.getTablePrivileges(identity.getUser(), tableName.getSchemaName(), tableName.getTableName()).stream()
+                .map(hivePrivilegeInfo -> hivePrivilegeInfo.getHivePrivilege().equals(toHivePrivilege(privilege)) ? hivePrivilegeInfo.isGrantOption() : false)
```
`hivePrivilegeInfo.getHivePrivilege().equals(toHivePrivilege(privilege)) ? hivePrivilegeInfo.isGrantOption() : false` should be replaced with `hivePrivilegeInfo.getHivePrivilege().equals(toHivePrivilege(privilege)) && hivePrivilegeInfo.isGrantOption()`

dain replied:
I'd use `.filter(hivePrivilegeInfo.getHivePrivilege().equals(toHivePrivilege(privilege))` and then just use `.anyMatch(HivePrivilege::isGrantOption))`

dain replied:
Actually, if we add `equals` and `hashcode` to HivePrivilegeInfo then we can just use `metastore.getTablePrivileges(...).contains(new HivePrivilegeInfo(required, true))`

amrutagokhale replied:
That's a good simplification.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java:
```diff
@@ -189,24 +190,36 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
+    {
+        return metastore.getTablePrivileges(identity.getUser(), tableName.getSchemaName(), tableName.getTableName()).stream()
+                .map(hivePrivilegeInfo -> hivePrivilegeInfo.getHivePrivilege().equals(toHivePrivilege(privilege)) ? hivePrivilegeInfo.isGrantOption() : false)
+                .anyMatch(b -> b == true);
```
Instead of `.map(...).anyMatch(b -> b == true)`, do `.anyMatch(...)`

amrutagokhale replied:
Given the simplification suggested by Dain, this goes away.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java:
```diff
@@ -189,24 +190,36 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
     {
-        Set<HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName);
+        Set<HivePrivilegeInfo.HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName).stream()
+                .map(hivePrivilegeInfo -> hivePrivilegeInfo.getHivePrivilege())
```
use method reference

amrutagokhale replied:
done

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/SqlStandardAccessControl.java:
```diff
@@ -189,24 +190,36 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
-        Set<HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName);
+        Set<HivePrivilegeInfo.HivePrivilege> privilegeSet = metastore.getDatabasePrivileges(identity.getUser(), schemaName).stream()
+                .map(hivePrivilegeInfo -> hivePrivilegeInfo.getHivePrivilege())
+                .collect(Collectors.toSet());
```
anyMatch

amrutagokhale replied:
Didn't quite get this. Can you elaborate?

dain commented:
@amrutagokhale, @haozhun will apply the changes directly and merge

amrutagokhale commented:
@haozhun I have a local patch ready that addresses the comments. I can push it to the remote and then you can take it from there? Let me know if that works.

haozhun commented:
Merged #4935.
